Enable more than Foodsoft format imports

robwa opened the pull request:
As a first step towards implementing #1100, the ability to import article data from formats other than the Foodsoft CSV format (e.g. BNN format) is required. The changes make it possible to select the format before uploading.

For this purpose, the existing library linked in #1100 is integrated into the code.

The following formats are supported:
- Foodsoft-CSV (ported from existing implementation)
- BNN
- ODIN

Bioromeo has a library implementation but no test data, so we do not offer this format for the time being.

The implementation is still work-in-progress, tests are still failing, etc.

RayOei commented:
> Bioromeo has a library implementation but no test dat

I have examples of BioRomeo data. We (FCAdamNoord) currently rework it manually with several regexp to get it to an importable format.
It would be nice to have this automated. The original import in the Adam fork also didn't work anymore for some years.
But, as far as I can see, BioRomeo is not very IT savvy and the article list which we get may not be very consistent over time.
I also understand we are the only one(?) using BioRomeo?
In short: the manual step is workable (I discussed internally).
I can create an issue for updating/implementing the BioRomeo part for reference & reminder?

RayOei commented:
Not sure who BNN is but as far as I know ODIN is used by many NL based foodcoops (except the one I am in). So I presume it is a feature that will be used. I can ask around, though, if that helps?

robwa commented:
> I noticed `roo` being downgraded to `2.9.0` from `2.10.1`. Any specific reason for that?

Good point! The Gemspec of the plugin originally contained the older `roo` version. I have fixed that.

RayOei commented:
I realised that this is being merged into master. I am not entirely sure (always have troubles finding the notes 😊) but I recall that in a recent dev meeting the proposal was to introduce a new `5.0 develop` branch and work from there, so the current master is only needed for fixes on the current `4.9.x` version. Albeit it would be better if even that would use dev branch to work from. Anyway...
In general I would suggest to work from a `develop` branch and branch `features` and what not from there. That makes it easier and safer to merge and work on until the moment it is safe to merge to`master` at some point in the future as a new release. As it will not influence the current production version, while developing.

RayOei commented:
Which also means I would be more inclined to `approve` as there is nothing that goes to `master`, at this point. As mentioned: I am not that confident that I don't miss important things as I am new to the code base and RoR/HAML for that matter. And for this specific feature I currently miss too much background and options (like real import files) to test with, which is what I plan to do as soon as I receive some ODIN files.

twothreenine commented on plugins/article_import/lib/foodsoft_article_import/bnn.rb:
```diff
@@ -0,0 +1,93 @@
+            availability: %w[X V].exclude?(row[1]),
+            name: UTF8Encoder.clean(row[6]),
+            order_number: row[0],
+            unit: row[21],
```
Custom units should be avoided preferrably. However, I see the benefit of importing the original unit string into some field, instead of simply setting supplier_order_unit to XPK (Package).

robwa replied:
Please see my answer above about field content reliability and manual adaption of units.

I don't agree that we want to avoid custom units in automatic import scenarios. From my point of view I would go for reliability first (combined with manual adaption). That is what the database unit migrations do as well.

For me the most important information on automatic import seems to be:
- container size (`row[22]`)
- price (`row[37]`)

All the rest might be nice to have, but only if it works well for all scenarios. Otherwise there could be a lot of confusion.

twothreenine commented on plugins/article_import/lib/foodsoft_article_import/bnn.rb:
```diff
@@ -0,0 +1,93 @@
+        unless row[0] == '' || row[0].nil?
+          article = {
+            availability: %w[X V].exclude?(row[1]),
+            name: UTF8Encoder.clean(row[6]),
```
If this was ok for now, we might leave it as it is, but:

The optional row[8] provides a cleaner, shorter article name without producer and unit. We could check if it is filled, and if not, use row[6] instead. However, there might be multiple articles with the same name in row[8], so we'd have to check for differences (producer, unit, origin, quality etc.) and specify them in brackets at the end of the article name, since articles of one supplier must have unique names in Foodsoft.
But it might also be useful to see the supplier directly in the article name.

If this is changed, `note` should be changed as well.

robwa replied:
In my examples, `row[8]` is unfortunately always empty.

The article name is also specified for automatic orders (see #1126). I am not sure whether the catalog name should appear in the order (the order number should of course always be the primary criterion).

Again, I see it the same way as before: When importing, information can be adjusted as desired and errors (such as rare duplicate names) can also be corrected manually. As a developer, I wouldn't like to make specifications to foodcoops how item names should be designed. I suspect that there are many different tastes.

twothreenine reviewed (commented):
I couldn't test it locally since I have problems with updating my Foodsoft clone in Docker. Do you have a hosted demo by any chance?

We've never used SharedLists in Austrian food coops nor am I familiar with BNN or any of the other formats. I looked at the BNN specification and sample data and made some suggestions, but I don't have any practical experience with it.

> I realised that this is being merged into master. I am not entirely sure (always have troubles finding the notes 😊) but I recall that in a recent dev meeting the proposal was to introduce a new 5.0 develop branch and work from there, so the current master is only needed for fixes on the current 4.9.x version. Albeit it would be better if even that would use dev branch to work from. 

We decided to use master as a dev branch and add a warning note. The whole new article unit system is still in dev stage. There's a [release branch](https://github.com/foodcoops/foodsoft/tree/v4.x) with the latest version before the article unit fork was merged.

robwa commented:
> I couldn't test it locally since I have problems with updating my Foodsoft clone in Docker. Do you have a hosted demo by any chance?

No, sorry, I only have my local development setup.

robwa commented on plugins/article_import/README.md:
```diff
@@ -0,0 +1,44 @@
+||4322Q|Tomato juice|Organic|Greenfields|TR|1.2 l|4.02|7.0|0|2|||Juices|
+
+bnn file reference:
+https://n-bnn.de/fileadmin/user_upload/Leistungen_Services/Markt-_und_Produktdaten/BNN3_1Schnittstelle_20.08.21.pdf
```
Good catch, I just copied the README from the original source. I replaced the link now as suggested. I am not sure if we want to maintain documentation at this place at all? Do we maybe want to remove the README completely?

robwa commented on app/views/articles/upload.html.haml:
There will be a format field in `Supplier` in some upcoming pull request but I am not sure if we want to use it in this case. There are already a lot of buttons in the interface (when remote synchronization is activated even two more) and another one feels like too much for me. I also don't have the feeling that selecting the file format when uploading a file would be a hassle. If the feature is used frequently, we could even recognise the format automatically in future. However, I suspect that, at least in the case of BNN, FTP synchronisation (not yet implemented here) will be used most of the time.

In this respect, I had already considered not offering file uploads at all. However, this would deprive users of a feature that is easy to implement.

The closest thing I could imagine is a dropdown, but I would suggest offering only one item (something like ‘Upload with format selection’) as an alternative. And I would parameterise the form so that we don't maintain the form code twice. But I'm not really happy with this solution either.

Any more opinions on this question?

twothreenine replied:
My main issue is with the missing Foodsoft format in-app documentation and having to make extra clicks before uploading a file.
There are ~70 food coops on our Austrian hosting, no SharedLists, but CSV upload seems to be popular. Either for editing articles as a spreadsheet (it's faster to download the CSV and edit them in a spreadsheet editor than editing them in-app one by one), or for using an external script which translates a custom format price list into a Foodsoft CSV (which we heavily use in my food coop).
That's why I'd like to keep it as straight-forward as possible.
Especially since the extended units come with big changes to the Foodsoft format.

How about keeping the Foodsoft format upload page as it was (incl. documentation), and adding a dropdown next to the button with a link to a separate upload page for the other formats?
> ![443853181-8d201534-46f7-4082-9779-4165ec1d2255](https://github.com/user-attachments/assets/4bc5a08f-0581-4517-a0e3-10baead96014)

twothreenine replied:
Ok, another suggestion:
- one form with a dropdown for the file formats, with Foodsoft format (csv/ods/xls/xlsx) as default
- depending on the selected format, a link to the format documentation is displayed
- for now, we put the Foodsoft format documentation in a separate view (in locales)
- when the external documentation is ready (in multiple languages), we can replace that link and remove the documentation view

Otherwise I see the risk that the external documentation won't be finished when these changes get released, and there are significant changes to the Foodsoft format, without any documentation.

kidhab replied:
> so there would be significant changes to the Foodsoft format, without any documentation.

That's a valid point. For now this view was the only place with documenation about the new Foodsoft import format.

robwa commented on plugins/article_import/lib/foodsoft_article_import/bnn.rb:
```diff
@@ -0,0 +1,93 @@
+            name: UTF8Encoder.clean(row[6]),
+            order_number: row[0],
+            unit: row[21],
+            article_unit_ratios: [{ sort: 1, quantity: row[22], unit: piece_unit_code }],
```
> I'd suggest parsing the value of row[23] with a regex into number and unit. If the number is different to 1, add a second ratio with quantity (number) and the respective unit code (KGM for kg etc., I think there's a translation method somewhere).

In a real world example the following values appear in `row[23]` ("Ladeneinheit"):
```
LT, FL, BE, PA, ST, GL, EI, KT, KG, SCH, DO, BTL, SC, BX, SET, VE, RO, KA, BOX, FA, LTR, T�, TU, SA
```
There are no numbers at all and I am not sure, if we can rely on the value being parsable or translatable. We could of course try to do so.

Unfortunately, there are also differences between documentation and actual use with regard to the other aspects. We cannot rely on the content of the fields. However, it is important to use `row[22]` (this information is reliable), as this results in the container size.

I assume that foodcoops will import the articles individually (in future via FTP). They will then have the option of adapting the order units to their requirements.

I also assume that we will continue to work on this feature as soon as the real-world applications are clearer. At the moment, I find it very difficult to keep track of all the possible scenarios. The whole thing is essentially supposed to be a simplification (automatic import from wholesale catalogs). If it doesn't work reliably or satisfactorily, there is always the fallback option of entering items manually.

I would therefore argue in favor of initially incorporating less unit magic and rather expanding it gradually and carefully on request.

Do you see any actual errors with the units?

twothreenine replied:
Can you send me some real world example files?

lentschi commented:
@robwa 
Sorry I'm late in the game - I haven't had time to read my foodsoft mails in the last few weeks.

Just one general question before I start reviewing:
Why are you pulling the source of https://git.local-it.org/Foodsoft/foodsoft_article_import into foodsoft itself? Wouldn't it be better to leave this in a separate repo? (I don't know too much about the internals of this, I'm just curious about your reasons for this decisions.)

robwa commented:
> Why are you pulling the source of https://git.local-it.org/Foodsoft/foodsoft_article_import into foodsoft itself? Wouldn't it be better to leave this in a separate repo?

I thought about it at the beginning and came to the conclusion that we probably don't want to maintain another separate repository. If it is different, we can of course move the code to a separate repository. But it would have to be one where we all can make adjustments (e.g. here on Github), as the code has to be changed.

lentschi commented:
> I chose the approach with the integration of the existing plugin because it was linked in the corresponding issue. I can also choose a different approach, but I would need clear indications of how you envisage this.

Well, since so far I've been the only one to criticize the direct integration, I'd say just go with how it's done in this PR :+1: (After all it can always be moved to a separate repo later, if the community finds it to cumbersome to maintain the imports.)

> I would appreciate a list of what needs to be changed so that I can merge this PR.

Well, as from my point of view, the list currently encompasses the following items:

- https://github.com/foodcoops/foodsoft/pull/1173#discussion_r2106230890
- **Optional** (I would still approve the PR, if not implemented!) general suggestions in https://github.com/foodcoops/foodsoft/pull/1173#issuecomment-2907886120
- **Post merge** https://github.com/foodcoops/foodsoft/pull/1173#discussion_r2106232892

I don't know about your review comments though, @twothreenine - maybe you could resolve those that you think are clarified or - if not - continue the respective discussions (robwa has already answered all of your requests)...?

@robwa Thanks again for taking your time to implement this!

robwa commented:
> * The file formats available for import in `plugins/article_import/lib/foodsoft_article_import/` should be referenced in detail. For example, the comments section of `bnn.rb` just refers to `www.n-bnn.de`, but something like "`https://n-bnn.de/leistungen-services/markt-und-produktdaten/schnittstelle` - API version 3.1" would be more precise.

I adapted the link and added a hint to reference in docs. At least for BNN.

> * Also we could consider copying English versions of the API descriptions to `doc/` (Otherwise, the linked API docs might just be altered at a later point and we'd lose the references)

Added at least the German specs to `doc/formats`.

RayOei commented on plugins/article_import/lib/foodsoft_article_import/bioromeo.rb:
Can you point me to it? So I can compare it with the latest XLS I have from them. I suspect it may not be aligned anymore because I understood from our (AdamNoord) member who does the article support, that the import doesn't work without manually changing it. But this would be a good starting point. I will wait for you to finish this and than take a stab at updating it (with a new issue) as soon as I have solved my Mollie problems 😁

robwa replied:
This is the file: https://github.com/foodcoops/foodsoft/blob/feature/1100-article-import/spec/plugins/article_import/files/bioromeo/bioromeo_flawless.csv

lentschi commented:
@robwa @yksflip @twothreenine @RayOei 
FYI: I just created a general discussion about the maintenance of plugins, which I think is related to this:
https://github.com/foodcoops/foodsoft/discussions/1193

steffen-unicode commented on Gemfile:
```diff
@@ -66,6 +66,7 @@ gem 'terser', '~> 1.1'
 
 # we use the git version of acts_as_versioned, and need to include it in this Gemfile
 gem 'acts_as_versioned', git: 'https://github.com/technoweenie/acts_as_versioned.git'
+gem 'foodsoft_article_import', path: 'plugins/article_import'
```
Should we disable this plugin by default according to the proposed [merge guidelines](https://github.com/foodcoops/foodsoft/discussions/1139#discussioncomment-13237875)?

lentschi replied:
Hm... After some discussion, I changed my opinion about that. As I think, [default plugins should be maintainable by everyone](https://github.com/foodcoops/foodsoft/discussions/1193#discussioncomment-13517503), this plugin isn't really eligible for that.

An idea: Could you maybe move the core lib (`FoodsoftArticleImport`) and the foodsoft import (`FoodsoftArticleImport`) from the plugin to the core and have the (then non-default) plugin include the rest of the import formats (all of them country-specific)?

twothreenine reviewed (changes requested):
I request the following changes (already described [here](https://github.com/foodcoops/foodsoft/pull/1173#discussion_r2129166278)):

> My main issue is with the missing Foodsoft format in-app documentation and having to make extra clicks before uploading a file.
> [...]
> How about keeping the Foodsoft format upload page as it was (incl. documentation), and adding a dropdown next to the button with a link to a separate upload page for the other formats (where you then have to select between BNN, ODIN etc.)?

RayOei commented:
I don't have a clear insights on the architecture of this plugin but it would make, logically, sense to have a core import functionality which can be expanded by a plugin. That way you can make a plugin `per` country and/or supplier. 
This would make maintenance also more transparant, having a plugin for `Bioremeo` and one for `Odin` (looking with the Dutch perspective) and so on for others.
